libr/lang/p/qjs/io.c

```diff
@@ -0,0 +1,258 @@
+					R_LOG_WARN ("exception2");
+					break;
+				}
+				buf[i] = u;
+			}
+			return len;
+		}
+		R_LOG_WARN ("return is not array");
+		return count;
+	}
+	// read stuff
+	return count;
+}
+
+static ut64 qjs_io_seek(RIO *io, RIODesc *fd, ut64 addr, int whence) {
+	RCore *core = io->coreb.core;
+	QjsPluginManager *pm = R_UNWRAP4 (core, lang, session, plugin_data);
+
+	// XXX this logic works for open or check, but not for seek, but its fine for 1 plugin for now
+	QjsIoPlugin *plugin;
+	R_VEC_FOREACH (&pm->io_plugins, plugin) {
+		JSContext *ctx = plugin->ctx;
+		JSValueConst args[2] = {
+			JS_NewInt64 (ctx, addr),
+			JS_NewInt32 (ctx, whence),
+		};
+		JSValue res = JS_Call (ctx, plugin->fn_seek_js, JS_UNDEFINED, countof (args), args);
+		int64_t radix = 10;
+		JS_ToInt64 (ctx, &radix, res);
+		return (ut64)radix;
+	}
+	return addr;
+}
+
+static RIODesc *qjs_io_open(RIO *io, const char *name, int perm, int mode) {
+	RCore *core = io->coreb.core;
+	QjsPluginManager *pm = R_UNWRAP4 (core, lang, session, plugin_data);
+
+	// Iterate over plugins until one returns "true" (meaning the plugin handled the input)
+	QjsIoPlugin *plugin;
+	R_VEC_FOREACH (&pm->io_plugins, plugin) {
+		JSContext *ctx = plugin->ctx;
+		JSValueConst args[3] = {
+			JS_NewString (ctx, name),
+			JS_NewInt32 (ctx, perm),
+			JS_NewInt32 (ctx, mode),
+	       	};
+		JSValue res = JS_Call (ctx, plugin->fn_open_js, JS_UNDEFINED, countof (args), args);
+		if (JS_ToBool (ctx, res)) {
+			RIOPlugin *iop = plugin->iop;
+			return r_io_desc_new (io, iop, name, perm, mode, core);
+		}
+	}
+
+	return NULL;
+}
+
+// TODO maybe add a function to call by plugin name? (is 1 extra arg)
+static char *qjs_io_system(RIO *io, RIODesc *fd, const char *input) {
+	RCore *core = io->coreb.core;
+	QjsPluginManager *pm = R_UNWRAP4 (core, lang, session, plugin_data);
+
+	// Iterate over plugins until one returns "true" (meaning the plugin handled the input)
+	QjsIoPlugin *plugin;
+	R_VEC_FOREACH (&pm->io_plugins, plugin) {
+		JSContext *ctx = plugin->ctx;
+		JSValueConst args[1] = { JS_NewString (ctx, input) };
+		JSValue res = JS_Call (ctx, plugin->fn_system_js, JS_UNDEFINED, countof (args), args);
+		if (JS_ToBool (ctx, res)) {
+			return strdup ("true");
+		}
+	}
+
+	return NULL;
+}
+
+static JSValue r2plugin_io(JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv) {
+	JSRuntime *rt = JS_GetRuntime (ctx);
+	QjsPluginManager *pm = JS_GetRuntimeOpaque (rt);
+
+	if (argc != 2) {
+		return JS_ThrowRangeError (ctx, "r2.plugin expects two arguments");
+	}
+
+	JSValueConst args[1] = { JS_NewString (ctx, ""), };
+
+	// check if res is an object
+	JSValue res = JS_Call (ctx, argv[1], JS_UNDEFINED, countof (args), args);
+	if (!JS_IsObject (res)) {
+		return JS_ThrowRangeError (ctx, "r2.plugin function must return an object");
+	}
+
+	JSValue name = JS_GetPropertyStr (ctx, res, "name");
+	size_t namelen;
+	const char *nameptr = JS_ToCStringLen2 (ctx, &namelen, name, false);
```

Comment:
This string is managed by qjs right? (In case we do an early return on an error later on..)